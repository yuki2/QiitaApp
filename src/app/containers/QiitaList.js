// @flow
import React, { Component } from 'react';
import { StyleSheet, View, FlatList } from 'react-native';
import _ from 'lodash';

import type { QiitaItem } from '../flow-type';

import QiitaCell from './QiitaCell';
import withIndicator from './withIndicator';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#FFFFFF',
  },
  listView: {
    flex: 1,
    backgroundColor: '#FFFFFF',
  },
  separator: {
    flex: 1,
    height: 1,
    backgroundColor: 'lightgray',
    marginLeft: 8,
  },
});

const separator = () => <View style={styles.separator} />;

type Props = {
  items: Array<QiitaItem>,
  loading: boolean,
  onSelectItem: (item: QiitaItem) => void,
  onRefresh: () => void,
  onEndReached: (distanceFromEnd: number, size: number) => void,
};

class QiitaList extends Component<Props> {
  static defaultProps = {
    items: [],
    loading: true,
    onSelectItem: () => {},
    onRefresh: () => {},
    onEndReached: () => {},
  };

  renderItem = ({ item }) => (
    <QiitaCell onSelect={() => this.props.onSelectItem(item)} item={item} />
  );

  renderListView = () => {
    const {
      loading, items, onRefresh, onEndReached,
    } = this.props;
    return (
      <View style={styles.container}>
        <FlatList
          initialNumToRender={10}
          style={styles.listView}
          data={items}
          renderItem={this.renderItem}
          keyExtractor={item => item.id}
          ItemSeparatorComponent={separator}
          refreshing={loading}
          onRefresh={onRefresh}
          onEndReachedThreshold={0.5}
          onEndReached={info => onEndReached(info.distanceFromEnd, _.size(items))}
        />
      </View>
    );
  };

  render = () => this.renderListView();
}

export default withIndicator(QiitaList);
