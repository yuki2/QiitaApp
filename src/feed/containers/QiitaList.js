import React, { Component } from 'react';
import { StyleSheet, View, FlatList } from 'react-native';
import PropTypes from 'prop-types';

import QiitaCell from './QiitaCell';
import QiitaIndicator from './QiitaIndicator';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#FFFFFF',
  },
  listView: {
    flex: 1,
    backgroundColor: '#FFFFFF',
  },
});

export default class QiitaList extends Component {
  static defaultProps = {
    items: [],
    loaded: false,
    onSelectItem: () => {},
  };

  selectItem = () => {};

  renderLoadingView = () => <QiitaIndicator />;

  renderItem = ({ item }) => (
    <QiitaCell onSelect={() => this.props.onSelectItem(item)} item={item} />
  );

  renderListView = () => (
    <View style={styles.container}>
      <FlatList
        style={styles.listView}
        data={this.props.items}
        renderItem={this.renderItem}
        keyExtractor={item => item.id}
      />
    </View>
  );

  render = () => {
    if (!this.props.loaded) {
      return this.renderLoadingView();
    }
    return this.renderListView();
  };
}

QiitaList.propTypes = {
  items: PropTypes.array,
  loaded: PropTypes.bool,
  onSelectItem: PropTypes.func,
};
