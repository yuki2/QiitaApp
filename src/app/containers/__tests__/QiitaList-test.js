import React from 'react';
import renderer from 'react-test-renderer';
import QiitaList from '../QiitaList';

const dummmyItem = (index) => {
  const user = {
    description: 'description',
    facebookId: 'facebookId',
    followeesCount: index,
    followersCount: index,
    githubLoginName: 'githubLoginName',
    id: `id${index}`,
    itemsCount: index,
    name: 'name',
    organization: 'organization',
    permanentId: 'permanentId',
    profileImageUrl: 'profileImageUrl',
    twitterScreenName: 'twitterScreenName',
    websiteUrl: 'websiteUrl',
  };
  return {
    id: `id${index}`,
    title: 'title',
    url: 'url',
    user,
    tags: ['tag1', 'tag2'],
    likesCount: index,
  };
};

it('renders correctly', () => {
  const items = [1, 2, 3, 4].map(index => dummmyItem(index));
  const tree = renderer.create(<QiitaList items={items} />).toJSON();
  expect(tree).toMatchSnapshot();
});
