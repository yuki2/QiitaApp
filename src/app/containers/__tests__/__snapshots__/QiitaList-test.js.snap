// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      Array [
        Object {
          "id": "id1",
          "likesCount": 1,
          "tags": Array [
            "tag1",
            "tag2",
          ],
          "title": "title",
          "url": "url",
          "user": Object {
            "description": "description",
            "facebookId": "facebookId",
            "followeesCount": 1,
            "followersCount": 1,
            "githubLoginName": "githubLoginName",
            "id": "id1",
            "itemsCount": 1,
            "name": "name",
            "organization": "organization",
            "permanentId": "permanentId",
            "profileImageUrl": "profileImageUrl",
            "twitterScreenName": "twitterScreenName",
            "websiteUrl": "websiteUrl",
          },
        },
        Object {
          "id": "id2",
          "likesCount": 2,
          "tags": Array [
            "tag1",
            "tag2",
          ],
          "title": "title",
          "url": "url",
          "user": Object {
            "description": "description",
            "facebookId": "facebookId",
            "followeesCount": 2,
            "followersCount": 2,
            "githubLoginName": "githubLoginName",
            "id": "id2",
            "itemsCount": 2,
            "name": "name",
            "organization": "organization",
            "permanentId": "permanentId",
            "profileImageUrl": "profileImageUrl",
            "twitterScreenName": "twitterScreenName",
            "websiteUrl": "websiteUrl",
          },
        },
        Object {
          "id": "id3",
          "likesCount": 3,
          "tags": Array [
            "tag1",
            "tag2",
          ],
          "title": "title",
          "url": "url",
          "user": Object {
            "description": "description",
            "facebookId": "facebookId",
            "followeesCount": 3,
            "followersCount": 3,
            "githubLoginName": "githubLoginName",
            "id": "id3",
            "itemsCount": 3,
            "name": "name",
            "organization": "organization",
            "permanentId": "permanentId",
            "profileImageUrl": "profileImageUrl",
            "twitterScreenName": "twitterScreenName",
            "websiteUrl": "websiteUrl",
          },
        },
        Object {
          "id": "id4",
          "likesCount": 4,
          "tags": Array [
            "tag1",
            "tag2",
          ],
          "title": "title",
          "url": "url",
          "user": Object {
            "description": "description",
            "facebookId": "facebookId",
            "followeesCount": 4,
            "followersCount": 4,
            "githubLoginName": "githubLoginName",
            "id": "id4",
            "itemsCount": 4,
            "name": "name",
            "organization": "organization",
            "permanentId": "permanentId",
            "profileImageUrl": "profileImageUrl",
            "twitterScreenName": "twitterScreenName",
            "websiteUrl": "websiteUrl",
          },
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.5}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControlMock
        onRefresh={[Function]}
        refreshing={true}
      />
    }
    refreshing={true}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <RCTRefreshControl />
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "backgroundColor": "white",
                "flex": 1,
                "paddingLeft": 8,
                "paddingRight": 8,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "marginBottom": 4,
                  "marginTop": 4,
                }
              }
            >
              <Image
                source={
                  Object {
                    "uri": "profileImageUrl",
                  }
                }
                style={
                  Object {
                    "height": 12,
                    "width": 12,
                  }
                }
              />
              <Text
                style={
                  Object {
                    "color": "gray",
                    "fontSize": 12,
                    "marginLeft": 4,
                    "textAlign": "left",
                  }
                }
              >
                id1
              </Text>
              <Text
                style={
                  Object {
                    "color": "gray",
                    "flex": 1,
                    "fontSize": 12,
                    "marginLeft": 4,
                    "textAlign": "right",
                  }
                }
              >
                a few seconds ago
              </Text>
            </View>
            <Text
              style={
                Object {
                  "fontSize": 15,
                  "marginBottom": 4,
                  "marginTop": 4,
                  "textAlign": "left",
                }
              }
            >
              title
            </Text>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "marginBottom": 4,
                  "marginTop": 4,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginBottom": 4,
                    "marginTop": 4,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "gray",
                        "fontSize": 12,
                      },
                      undefined,
                      Object {
                        "fontFamily": "Ionicons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                    ]
                  }
                >
                  
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "gray",
                        "fontSize": 12,
                        "marginLeft": 4,
                        "textAlign": "left",
                      },
                      Object {
                        "marginRight": 8,
                      },
                    ]
                  }
                >
                  1
                </Text>
              </View>
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "gray",
                      "fontSize": 12,
                    },
                    undefined,
                    Object {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  Object {
                    "color": "gray",
                    "fontSize": 12,
                    "marginLeft": 4,
                    "textAlign": "left",
                  }
                }
              >
                , 
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "lightgray",
              "flex": 1,
              "height": 1,
              "marginLeft": 8,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "backgroundColor": "white",
                "flex": 1,
                "paddingLeft": 8,
                "paddingRight": 8,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "marginBottom": 4,
                  "marginTop": 4,
                }
              }
            >
              <Image
                source={
                  Object {
                    "uri": "profileImageUrl",
                  }
                }
                style={
                  Object {
                    "height": 12,
                    "width": 12,
                  }
                }
              />
              <Text
                style={
                  Object {
                    "color": "gray",
                    "fontSize": 12,
                    "marginLeft": 4,
                    "textAlign": "left",
                  }
                }
              >
                id2
              </Text>
              <Text
                style={
                  Object {
                    "color": "gray",
                    "flex": 1,
                    "fontSize": 12,
                    "marginLeft": 4,
                    "textAlign": "right",
                  }
                }
              >
                a few seconds ago
              </Text>
            </View>
            <Text
              style={
                Object {
                  "fontSize": 15,
                  "marginBottom": 4,
                  "marginTop": 4,
                  "textAlign": "left",
                }
              }
            >
              title
            </Text>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "marginBottom": 4,
                  "marginTop": 4,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginBottom": 4,
                    "marginTop": 4,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "gray",
                        "fontSize": 12,
                      },
                      undefined,
                      Object {
                        "fontFamily": "Ionicons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                    ]
                  }
                >
                  
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "gray",
                        "fontSize": 12,
                        "marginLeft": 4,
                        "textAlign": "left",
                      },
                      Object {
                        "marginRight": 8,
                      },
                    ]
                  }
                >
                  2
                </Text>
              </View>
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "gray",
                      "fontSize": 12,
                    },
                    undefined,
                    Object {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  Object {
                    "color": "gray",
                    "fontSize": 12,
                    "marginLeft": 4,
                    "textAlign": "left",
                  }
                }
              >
                , 
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "lightgray",
              "flex": 1,
              "height": 1,
              "marginLeft": 8,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "backgroundColor": "white",
                "flex": 1,
                "paddingLeft": 8,
                "paddingRight": 8,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "marginBottom": 4,
                  "marginTop": 4,
                }
              }
            >
              <Image
                source={
                  Object {
                    "uri": "profileImageUrl",
                  }
                }
                style={
                  Object {
                    "height": 12,
                    "width": 12,
                  }
                }
              />
              <Text
                style={
                  Object {
                    "color": "gray",
                    "fontSize": 12,
                    "marginLeft": 4,
                    "textAlign": "left",
                  }
                }
              >
                id3
              </Text>
              <Text
                style={
                  Object {
                    "color": "gray",
                    "flex": 1,
                    "fontSize": 12,
                    "marginLeft": 4,
                    "textAlign": "right",
                  }
                }
              >
                a few seconds ago
              </Text>
            </View>
            <Text
              style={
                Object {
                  "fontSize": 15,
                  "marginBottom": 4,
                  "marginTop": 4,
                  "textAlign": "left",
                }
              }
            >
              title
            </Text>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "marginBottom": 4,
                  "marginTop": 4,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginBottom": 4,
                    "marginTop": 4,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "gray",
                        "fontSize": 12,
                      },
                      undefined,
                      Object {
                        "fontFamily": "Ionicons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                    ]
                  }
                >
                  
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "gray",
                        "fontSize": 12,
                        "marginLeft": 4,
                        "textAlign": "left",
                      },
                      Object {
                        "marginRight": 8,
                      },
                    ]
                  }
                >
                  3
                </Text>
              </View>
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "gray",
                      "fontSize": 12,
                    },
                    undefined,
                    Object {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  Object {
                    "color": "gray",
                    "fontSize": 12,
                    "marginLeft": 4,
                    "textAlign": "left",
                  }
                }
              >
                , 
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "lightgray",
              "flex": 1,
              "height": 1,
              "marginLeft": 8,
            }
          }
        />
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "backgroundColor": "white",
                "flex": 1,
                "paddingLeft": 8,
                "paddingRight": 8,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "marginBottom": 4,
                  "marginTop": 4,
                }
              }
            >
              <Image
                source={
                  Object {
                    "uri": "profileImageUrl",
                  }
                }
                style={
                  Object {
                    "height": 12,
                    "width": 12,
                  }
                }
              />
              <Text
                style={
                  Object {
                    "color": "gray",
                    "fontSize": 12,
                    "marginLeft": 4,
                    "textAlign": "left",
                  }
                }
              >
                id4
              </Text>
              <Text
                style={
                  Object {
                    "color": "gray",
                    "flex": 1,
                    "fontSize": 12,
                    "marginLeft": 4,
                    "textAlign": "right",
                  }
                }
              >
                a few seconds ago
              </Text>
            </View>
            <Text
              style={
                Object {
                  "fontSize": 15,
                  "marginBottom": 4,
                  "marginTop": 4,
                  "textAlign": "left",
                }
              }
            >
              title
            </Text>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "marginBottom": 4,
                  "marginTop": 4,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginBottom": 4,
                    "marginTop": 4,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": "gray",
                        "fontSize": 12,
                      },
                      undefined,
                      Object {
                        "fontFamily": "Ionicons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                    ]
                  }
                >
                  
                </Text>
                <Text
                  style={
                    Array [
                      Object {
                        "color": "gray",
                        "fontSize": 12,
                        "marginLeft": 4,
                        "textAlign": "left",
                      },
                      Object {
                        "marginRight": 8,
                      },
                    ]
                  }
                >
                  4
                </Text>
              </View>
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "gray",
                      "fontSize": 12,
                    },
                    undefined,
                    Object {
                      "fontFamily": "Ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  Object {
                    "color": "gray",
                    "fontSize": 12,
                    "marginLeft": 4,
                    "textAlign": "left",
                  }
                }
              >
                , 
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
